[ConstantFolding] Add ilogb in isMathLibCallNoop

kshitijvp opened the pull request:
ilogb libcall was not being constant folded correctly. This patch adds ilogb case in
isMathLibCallNoop with correct error condition. 

Fixes #101873

kshitijvp committed:
[LibCallShrinkWrap] Added ilogb in performCallErrors()

ilogb libcall was not being constant folded correctly.
It was due to LibCallShrinkWrap pass not handling
ilogbx, as a result SimplifyCFG pass did not remove
redundant libcall. This patch adds ilogb case in
performCallErrors().

Fixes #101873

kshitijvp commented:
@dtcxzyw @c8ef  I tried this approach and it seems to work, could you please review my PR. Thanks!

dtcxzyw reviewed (changes requested):
As @c8ef pointed out, we should fix it in `llvm::isMathLibCallNoop`. Constant folding is correct. But we have to prove that no errors occur in the call to eliminate the call to ilogb (See also https://en.cppreference.com/w/cpp/numeric/math/ilogb).

BTW, please pre-commit tests to demonstrate that your patch fixes the issue.
See also https://llvm.org/docs/InstCombineContributorGuide.html#precommit-tests.

VedantParanjape commented:
> And I am suspicious of the original code, which states that the error in calling `ilogb` is the same as `log`. However, according to the man pages, this is not true.
> 
> `ilogb`:
> 
> ```
> ERRORS         [top](https://man7.org/linux/man-pages/man3/ilogb.3.html#top_of_page)
>        See [math_error(7)](https://man7.org/linux/man-pages/man7/math_error.7.html) for information on how to determine whether an
>        error has occurred when calling these functions.
> 
>        The following errors can occur:
> 
>        Domain error: x is 0 or a NaN
>               An invalid floating-point exception (FE_INVALID) is
>               raised, and [errno](https://man7.org/linux/man-pages/man3/errno.3.html) is set to EDOM (but see BUGS).
> 
>        Domain error: x is an infinity
>               An invalid floating-point exception (FE_INVALID) is
>               raised, and [errno](https://man7.org/linux/man-pages/man3/errno.3.html) is set to EDOM (but see BUGS).
> ```
> 
> `log`:
> 
> ```
> See [math_error(7)](https://man7.org/linux/man-pages/man7/math_error.7.html) for information on how to determine whether an
>        error has occurred when calling these functions.
> 
>        The following errors can occur:
> 
>        Domain error: x is negative
>               [errno](https://man7.org/linux/man-pages/man3/errno.3.html) is set to EDOM.  An invalid floating-point exception
>               (FE_INVALID) is raised.
> 
>        Pole error: x is zero
>               [errno](https://man7.org/linux/man-pages/man3/errno.3.html) is set to ERANGE.  A divide-by-zero floating-point
>               exception (FE_DIVBYZERO) is raised.
> ```

looking at isMathLibCallNoop, I think we need checks in there as well as LibCallShrinkWrap. Since there are three Domain errors to be covered, right?

kshitijvp committed:
[ConstantFolding] Added ilogb in isMathLibCallNoop()

ilogb libcall was not being constant folded correctly.
This patch adds ilogb with its correct domain in isMathLibCallNoop().

Fixes #101873

kshitijvp commented:
@c8ef  @dtcxzyw i have added the required changes could you please review it 
Thanks!

kshitijvp committed:
added more tests

dtcxzyw commented on llvm/test/Transforms/Util/libcalls-shrinkwrap-float.ll:
Please drop this unrelated change. I would like to see that calls to ilogb are eliminated in the test `llvm/test/Transforms/InstCombine/ilogb.ll`.

VedantParanjape replied:
It seems this simplification is happening in EarlyCSE and not instcombine.

```
; *** IR Dump Before EarlyCSEPass on ilogb_const1 ***
define i32 @ilogb_const1() {
  %r = call i32 @ilogb(double -7.000000e+00)
  ret i32 %r
}
; *** IR Dump After EarlyCSEPass on ilogb_const1 ***
define i32 @ilogb_const1() {
  ret i32 2
}
```

kshitijvp replied:
Ping! @dtcxzyw

arsenm replied:
I recall debugging the same thing before. It's an oddity we don't delete the call here, but if we want to fix that it's beyond the scope of this patch

VedantParanjape replied:
> > It seems this simplification is happening in EarlyCSE and not instcombine. Ran this after applying OP's patch.
> > `opt -O3 llvm/test/Transforms/InstCombine/ilogb.ll -S -print-after-all -print-before-all -filter-print-funcs=ilogb_const1`
> > ```
> > ; *** IR Dump Before EarlyCSEPass on ilogb_const1 ***
> > define i32 @ilogb_const1() {
> >   %r = call i32 @ilogb(double -7.000000e+00)
> >   ret i32 %r
> > }
> > ; *** IR Dump After EarlyCSEPass on ilogb_const1 ***
> > define i32 @ilogb_const1() {
> >   ret i32 2
> > }
> > ```
> 
> The constant folding part is invoked in the EarlyCSEPass, but it does not delete the original call. As far as I know, to delete the original function call, we need to prove it has no side effects. This part happens in the SimplifyCFG pass, which utilizes the InstSimplify functionality and eventually calls llvm::isMathLibCallNoop.
> 
> ref: https://godbolt.org/z/azosfrK49

yes that's right, I had posted the same in the issue, so where should the testcase go for ilogb?

dtcxzyw commented on llvm/lib/Analysis/ConstantFolding.cpp:
```diff
@@ -3807,6 +3807,9 @@ bool llvm::isMathLibCallNoop(const CallBase *Call,
         return Op.isNaN() || (!Op.isZero() && !Op.isNegative());
 
+      case LibFunc_ilogb:
+        return !Op.isNaN() && (!Op.isZero() && !Op.isInfinity());
```
```suggestion
        return !Op.isNaN() && !Op.isZero() && !Op.isInfinity();
```

arsenm commented on llvm/test/Transforms/Util/libcalls-shrinkwrap-float.ll:
```diff
@@ -223,6 +223,8 @@ entry:
 ; CHECK-NEXT: br label %[[END_LABEL]]
 ; CHECK: [[END_LABEL]]:
+
+  %call_12 = call i32 @ilogb(float %value)
```
I don't understand why you are are adding new calls to existing tests, and not new test functions

kshitijvp replied:
sorry pls ignore this , it was to be removed

kshitijvp committed:
made changes to test file

kshitijvp committed:
fixed newline

kshitijvp commented:
I have fixed and added the test case.

arsenm commented on llvm/test/Transforms/InstSimplify/pr122582.ll:
```diff
@@ -0,0 +1,13 @@
+; NOTE: Assertions have been autogenerated by utils/update_test_checks.py UTC_ARGS: --version 5
+; RUN: opt < %s -O1 -S | FileCheck %s
```
```suggestion
; RUN: opt -S -passes=instsimplify < %s | FileCheck %s
```

Pass tests should run the minimum set of required passes

arsenm commented on llvm/test/Transforms/InstSimplify/pr122582.ll:
```diff
@@ -0,0 +1,13 @@
+  %r = call i32 @ilogb(double -7.000000e+00)
+  ret i32 %r
+}
+
```
Test a strictfp case, and all relevant FP types

arsenm replied:
Plus test all the special cases, particularly the ones that need to be skipped

arsenm commented on llvm/test/Transforms/InstSimplify/pr122582.ll:
```diff
@@ -0,0 +1,13 @@
+; RUN: opt < %s -O1 -S | FileCheck %s
+
+define i32 @ilogb_const1() {
+; CHECK-LABEL: define noundef i32 @ilogb_const1(
+; CHECK-SAME: ) local_unnamed_addr #[[ATTR0:[0-9]+]] {
+; CHECK-NEXT:    ret i32 2
```
Attribute noise will disappear with this reduced to running instsimplify

kshitijvp committed:
added strictfp tests

kshitijvp committed:
minor change

kshitijvp commented:
Made some minor changes. Can you review it and give me a lgtm.

kshitijvp committed:
removed whitespace

Co-authored-by: c8ef <c8ef@outlook.com>

kshitijvp commented on llvm/test/Transforms/InstSimplify/pr122582.ll:
```diff
@@ -0,0 +1,213 @@
+; CHECK-NEXT:    ret i32 2
+;
+  %r = call i32 @ilogb(double -7.000000e+00)
+  ret i32 %r      
```
Done.

arsenm reviewed (approved):
LGTM. We probably should use an explicit triple in the test though (I have a WIP patch to switch the default handling to assume no libcalls are available without a triple)

kshitijvp commented:
Thanks, I will surely make this change in another patch, can you share your WIP implementation if possible? I want to make more contributions and learn more :)

arsenm commented:
> Thanks, I will surely make this change in another patch, can you share your WIP implementation if possible? I want to make more contributions and learn more :)

It's not really in a postable state (and I kind of forgot about, it's many months old)

kshitijvp commented:
Okay, If you restart working on it, I'll be happy to contribute thanks!

It seems @dtcxzyw review is blocking this? I did address his requested change.

VedantParanjape commented:
> Since the patch is approved can someone merge it. Thanks!

Sure, merged! Thanks for your contribution.
